demos/classification_demo/main.cpp

```diff
@@ -0,0 +1,506 @@
+                }
+
+                framesNum += FLAGS_b;
+                
+                avgFPS = framesNum / std::chrono::duration_cast<Sec>(
+                    std::chrono::steady_clock::now() - startTickCount).count();
+                gridMat.updateMat(shownImagesInfo);
+                auto processingEndTime = std::chrono::steady_clock::now();
+                for (const auto & image : completedInferRequestInfo.images) {
+                    latencySum += processingEndTime - image.startTime;
+                }
+                avgLatency = std::chrono::duration_cast<Sec>(latencySum).count() / framesNum;
+                accuracy = static_cast<double>(correctPredictionsCount) / framesNum;
+                gridMat.textUpdate(avgFPS, avgLatency, accuracy, isTestMode, !FLAGS_gt.empty(), presenter);
+                
+                if (!FLAGS_no_show) {
+                    cv::imshow("classification_demo", gridMat.outImg);
+                    key = static_cast<char>(cv::waitKey(1));
+                    presenter.handleKey(key);
+                }
+            } else if (!emptyInferRequests.empty()) {
+                lock.unlock();
+                auto inferRequestStartTime = std::chrono::steady_clock::now();
+                cv::Mat nextImage = resizeImage(inputImages[nextImageIndex], modelInputResolution);
+                emptyInferRequests.front().images.push_back(
+                                                    {nextImage,
+                                                     classIndices[nextImageIndex],
+                                                     inferRequestStartTime});
+                nextImageIndex++;
+                if (nextImageIndex == inputImagesCount) {
+                    nextImageIndex = 0;
+                }
+                if (emptyInferRequests.front().images.size() == FLAGS_b) {
+                    auto emptyInferRequest = emptyInferRequests.front();
+                    emptyInferRequests.pop();
+
+                    emptyInferRequest.inferRequest.SetCompletionCallback([emptyInferRequest,
+                                                      &completedInferRequests,
+                                                      &mutex,
+                                                      &condVar,
+                                                      &hasCallbackException,
+                                                      &irCallbackException] {
+                        try {
+                            std::lock_guard<std::mutex> callback_lock(mutex);
+                            completedInferRequests.push({emptyInferRequest.inferRequest, emptyInferRequest.images});
+                        }
+                        catch(...) {
+                            if (!hasCallbackException) {
+                                irCallbackException = std::current_exception();
+                                hasCallbackException = true;
+                            }
```

Comment:
I don't think this is thread safe. If multiple `InferRequest`s run their callbacks concurrently, then they can both see `hasCallbackException` as false, and attempt to reset `irCallbackException` at the same time.